sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt

```diff
@@ -0,0 +1,70 @@
+package com.infinum.sentinel.ui.certificates.observer
+
+import android.content.Context
+import android.util.Log
+import androidx.annotation.UiThread
+import androidx.work.Data
+import androidx.work.ListenableWorker
+import androidx.work.WorkerFactory
+import androidx.work.WorkerParameters
+import com.google.common.util.concurrent.ListenableFuture
+import com.infinum.sentinel.ui.shared.Constants.Keys.WORKER_CLASS_NAME
+import com.infinum.sentinel.ui.shared.Constants.Keys.WORKER_ID
+import java.util.concurrent.Executor
+import java.util.concurrent.TimeUnit
+
+/**
+ * A worker to delegate work requests from within the library to workers
+ * that require factories with custom dependencies.
+ */
+internal class DelegateWorker(
+    appContext: Context,
+    parameters: WorkerParameters,
+) : ListenableWorker(appContext, parameters) {
+
+    private val workerClassName =
+        parameters.inputData.getString(WORKER_CLASS_NAME) ?: ""
+    private val workerId = parameters.inputData.getString(WORKER_ID)
+    private val delegateWorkerFactory = workerFactories[workerId]
+    private val delegatedWorker = delegateWorkerFactory?.createWorker(appContext, workerClassName, parameters)
+
+    override fun startWork(): ListenableFuture<Result> {
+        return if (delegatedWorker != null) {
+            delegatedWorker.startWork()
+        } else {
+            val errorMessage = "No delegateWorker available for $workerId" +
+                " with workerClassName of $workerClassName. Is the " +
+                "RouterWorker.workerFactories populated correctly?"
+
+            Log.w("Sentinel", errorMessage)
+
+            val errorData = Data.Builder().putString("Reason", errorMessage).build()
+
+            object : ListenableFuture<Result> {
+                override fun isDone(): Boolean = true
+                override fun get(): Result = Result.failure(errorData)
+                override fun get(timeout: Long, unit: TimeUnit): Result = Result.failure(errorData)
+                override fun cancel(mayInterruptIfRunning: Boolean): Boolean = false
+                override fun isCancelled(): Boolean = false
+                override fun addListener(listener: Runnable, executor: Executor) = listener.run()
+            }
```

Comment:
Why do we return an "empty" future instead of throwing an error? Shouldn't throwing an error be safer?

Reply:
Well, it is not "empty". It is failure with errorData. We need to return `ListenableFuture<Result>`

Reply:
Yeah, but I'm exploring less "silent" ways of this failing. Realistically, it could happen only if we forget to register a worker factory, but that's something we shouldn't allow to end up in a release. I'm trying to figure out how we could make sure to not fail fast and loud :D

Reply:
Do you think just exception will do the job?

Reply:
I'm open for discussions, but an exception (if it would crash the app) would definitely be more noticeable.

Reply:
If we throw exception in this place it will just produce logs, similar to Result.Failure. Because it is running on background thread 
```
Work [ id=42358918-83db-4c07-847b-6020296c3d41, tags={ com.infinum.sentinel.ui.certificates.observer.DelegateWorker } ] failed because it threw an exception/error
                                                                                                    java.util.concurrent.ExecutionException: java.lang.IllegalStateException: No delegateWorker available for com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker with workerClassName of com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker. Is the DelegateWorker.workerFactories populated correctly?
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:515)
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
                                                                                                    	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
                                                                                                    	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    Caused by: java.lang.IllegalStateException: No delegateWorker available for com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker with workerClassName of com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker. Is the RouterWorker.workerFactories populated correctly?
                                                                                                    	at com.infinum.sentinel.ui.certificates.observer.DelegateWorker.startWork(DelegateWorker.kt:40)
                                                                                                    	at androidx.work.impl.WorkerWrapper$1.run(WorkerWrapper.java:302)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7872)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
2024-07-17 16:57:47.222 15350-15401 WM-WorkerWrapper        com.infinum.sentinel.sample.debug    I  Worker result FAILURE for Work [ id=42358918-83db-4c07-847b-6020296c3d41, tags={ com.infinum.sentinel.ui.certificates.observer.DelegateWorker } ] 
```

But error is defiantly more noticeable in logs

Reply:
If things go 💥 at least people will see errors and open issues 😅